src/usr.bin/sudo/visudo.8
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
''' Revision 1.4  2000/03/27 03:26:24  millert
''' Use 8 and 5 in the man page bodies as well.
'''
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
'''   \*(M", \*(S", \*(N" and \*(T" are the equivalent of
'''   \*(L" and \*(R", except that they are used on ".xx" lines,
'''   such as .IP and .SH, which do another additional levels of
'''   double-quote interpretation
.ds M" """
.ds S" """
.ds N" """""
.ds T" """""
.ds L' '
.ds R' '
.ds M' '
.ds S' '
.ds N' '
.ds T' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds M" ``
.ds S" ''
.ds N" ``
.ds T" ''
.ds L' `
.ds R' '
.ds M' `
.ds S' '
.ds N' `
.ds T' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH visudo 8 "1.6.3" "26/Mar/2000" "MAINTENANCE COMMANDS"
.UC
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
visudo \- edit the sudoers file
.SH "SYNOPSIS"
\fBvisudo\fR [ \fB\-s\fR ] [ \fB\-V\fR ]
.SH "DESCRIPTION"
\fBvisudo\fR edits the \fIsudoers\fR file in a safe fashion, analogous to
\fIvipw\fR\|(8).  \fBvisudo\fR locks the \fIsudoers\fR file against multiple
simultaneous edits, provides basic sanity checks, and checks
for parse errors.  If the \fIsudoers\fR file is currently being
edited you will receive a message to try again later.  In the
default configuration, the \fIvi\fR\|(1) editor is used, but there is
a compile time option to allow use of whatever editor the
environment variables \f(CWEDITOR\fR or \f(CWVISUAL\fR are set to.
.PP
\fBvisudo\fR parses the \fIsudoers\fR file after the edit and will
not save the changes if there is a syntax error.  Upon finding
an error, a message will be printed stating the line \fInumber\fR\|(s)
that the error occurred on and the user will receive the
\*(L"What now?\*(R" prompt.  At this point the user may enter \*(L"e\*(R"
to re-edit the \fIsudoers\fR file, enter \*(L"x\*(R" to exit without
saving the changes, or \*(L"Q\*(R" to quit and save changes.  The
\*(L"Q\*(R" option should be used with extreme care because if \fBvisudo\fR
believes there to be a parse error, so will \fBsudo\fR and no one
will be able to execute \fBsudo\fR again until the error is fixed.
Any other command at this prompt will print a short help message.
When editing the \fIsudoers\fR file after a parse error has been
detected the cursor will be placed on the line where the error
occurred (if the editor supports this feature).
.SH "OPTIONS"
\fBvisudo\fR accepts the following command line option:
.Ip "-s" 4
Enable \fBstrict\fR checking of the \fIsudoers\fR file.  If an alias is
used before it is defined, \fBvisudo\fR will consider this a parse
error.  Note that it is not possible to differentiate between an
alias and a hostname or username that consists solely of upper case
letters, digits, and the underscore ('_') character.
.Ip "-V" 4
The \f(CW-V\fR (version) option causes \fBvisudo\fR to print the version number
and exit.
.SH "ERRORS"
.Ip "sudoers file busy, try again later." 4
Someone else is currently editing the \fIsudoers\fR file.
.Ip "/etc/sudoers.tmp: Permission denied" 4
You didn't run \fBvisudo\fR as root.
.Ip "Can't find you in the passwd database" 4
Your userid does not appear in the system passwd file.
.Ip "Warning: undeclared Alias referenced near ..." 4
Either you are using a {User,Runas,Host,Cmnd}_Alias before
defining it or you have a user or hostname listed that
consists solely of upper case letters, digits, and the
underscore ('_') character.  If the latter, you can ignore
the warnings (\fBsudo\fR will not complain).  In \fB\-s\fR (strict)
mode these are errors not warnings.
.SH "ENVIRONMENT"
The following environment variables are used only if \fBvisudo\fR
was configured with the \fI--with-env-editor\fR option:
.PP
.Vb 2
\& EDITOR                 Used by visudo as the editor to use
\& VISUAL                 Used by visudo if EDITOR is not set
.Ve
.SH "FILES"
.PP
.Vb 2
\& /etc/sudoers           List of who can run what
\& /etc/sudoers.tmp       Lock file for visudo
.Ve
.SH "AUTHOR"
Many people have worked on \fIsudo\fR over the years, this version of
\fBvisudo\fR was written by:
.PP
.Vb 1
\& Todd Miller            <Todd.Miller@courtesan.com>
.Ve
See the HISTORY file in the sudo distribution for more details.
.SH "BUGS"
If you feel you have found a bug in sudo, please submit a bug report
at http://www.courtesan.com/sudo/bugs/
.SH "DISCLAIMER"
\fBVisudo\fR is provided ``AS IS'\*(R' and any express or implied warranties,
including, but not limited to, the implied warranties of merchantability
and fitness for a particular purpose are disclaimed.
See the LICENSE file distributed with \fBsudo\fR for complete details.
.SH "CAVEATS"
There is no easy way to prevent a user from gaining a root shell if 
the editor used by \fBvisudo\fR allows shell escapes.
.SH "SEE ALSO"
\fIsudo\fR\|(8), \fIvipw\fR\|(8).

.rn }` ''
.IX Title "visudo 8"
.IX Name "visudo - edit the sudoers file"

.IX Header "NAME"

.IX Header "SYNOPSIS"

.IX Header "DESCRIPTION"

.IX Header "OPTIONS"

.IX Item "-s"

.IX Item "-V"

.IX Header "ERRORS"

.IX Item "sudoers file busy, try again later."

.IX Item "/etc/sudoers.tmp: Permission denied"

.IX Item "Can't find you in the passwd database"

.IX Item "Warning: undeclared Alias referenced near ..."

.IX Header "ENVIRONMENT"

.IX Header "FILES"

.IX Header "AUTHOR"

.IX Header "BUGS"

.IX Header "DISCLAIMER"

.IX Header "CAVEATS"

.IX Header "SEE ALSO"



src/gnu/usr.bin/sudo/visudo/visudo.c
/*	$OpenBSD: visudo.c,v 1.6 1998/01/13 05:30:32 millert Exp $	*/

/*
 *  CU sudo version 1.5.5
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 1, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *  Please send bugs, changes, problems to sudo-bugs@courtesan.com
 *
 *******************************************************************
 *
 *  visudo.c -- locks the sudoers file for safe editing and check
 *  for parse errors.
 *
 *  Todd C. Miller (millert@colorado.edu) Sat Mar 25 21:50:36 MST 1995
 */

#ifndef lint
static char rcsid[] = "$Id$";
#endif /* lint */

#include "config.h"

#include <stdio.h>
#ifdef STDC_HEADERS
#include <stdlib.h>
#endif /* STDC_HEADERS */
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif /* HAVE_UNISTD_H */
#ifdef HAVE_STRING_H
#include <string.h>
#endif /* HAVE_STRING_H */
#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif /* HAVE_STRINGS_H */
#if defined(HAVE_MALLOC_H) && !defined(STDC_HEADERS)
#include <malloc.h>
#endif /* HAVE_MALLOC_H && !STDC_HEADERS */
#include <ctype.h>
#include <pwd.h>
#include <signal.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <netinet/in.h>

#include "sudo.h"
#include <options.h>
#include "version.h"

#ifndef STDC_HEADERS
#ifndef __GNUC__	/* gcc has its own malloc */
extern char *malloc	__P((size_t));
#endif /* __GNUC__ */
extern char *getenv	__P((const char *));
extern int stat		__P((const char *, struct stat *));
#endif /* !STDC_HEADERS */

#if defined(POSIX_SIGNALS) && !defined(SA_RESETHAND)
#define SA_RESETHAND    0
#endif /* POSIX_SIGNALS && !SA_RESETHAND */

/*
 * Function prototypes
 */
static void usage		__P((void));
static char whatnow		__P((void));
static void whatnow_help	__P((void));
static RETSIGTYPE Exit		__P((int));
static void setup_signals	__P((void));
int command_matches		__P((char *, char *, char *, char *));
int addr_matches		__P((char *));
int netgr_matches		__P((char *, char *, char *));
int usergr_matches		__P((char *, char *));
void init_parser		__P((void));


/*
 * External globals
 */
extern FILE *yyin, *yyout;
extern int errorlineno, sudolineno;


/*
 * Globals
 */
char **Argv;
char **NewArgv = NULL;
int NewArgc = 0;
char *sudoers = _PATH_SUDO_SUDOERS;
char *stmp = _PATH_SUDO_STMP;
int parse_error = FALSE;
char *runas_user = "root";

/*
 * For the parsing routines
 */
char host[] = "";
char *shost = "";
char *cmnd = "";
char *cmnd_args = NULL;
struct passwd *user_pw_ent;


/********************************************************************
 *
 *  main()
 *
 *  where it all begins...
 */

int main(argc, argv)
    int argc;
    char **argv;
{
    char buf[MAXPATHLEN*2];		/* buffer used for copying files */
    char * Editor = EDITOR;		/* editor to use (default is EDITOR */
    int sudoers_fd;			/* sudoers file descriptor */
    int stmp_fd;			/* stmp file descriptor */
    int n;				/* length parameter */

    (void) setbuf(stderr, (char *)NULL);	/* unbuffered stderr */

    /*
     * Parse command line options
     */
    Argv = argv;

    /*
     * If passesd -V then print version, else print usage
     * if any other option...
     */
    if (argc == 2)
	if (!strcmp(Argv[1], "-V")) {
	    (void) printf("visudo version %s\n", version);
	    exit(0);
	} else {
	    usage();
	}
    else if (argc != 1)
	usage();

    /* user_pw_ent needs to point to something... */
    if ((user_pw_ent = getpwuid(getuid())) == NULL) {
	(void) fprintf(stderr, "%s: Can't find you in the passwd database: ",
	    Argv[0]);
	perror(stmp);
	exit(1);
    }

#ifdef ENV_EDITOR
    /*
     * If we are allowing EDITOR and VISUAL envariables set Editor
     * base on whichever exists...
     */
    if (!(Editor = getenv("EDITOR")))
	if (!(Editor = getenv("VISUAL")))
	    Editor = EDITOR;
#endif /* ENV_EDITOR */

    /*
     * Copy sudoers file to stmp
     */
    stmp_fd = open(stmp, O_WRONLY | O_CREAT | O_EXCL, 0600);
    if (stmp_fd < 0) {
	if (errno == EEXIST) {
	    (void) fprintf(stderr, "%s: sudoers file busy, try again later.\n",
		Argv[0]);
	    exit(1);
	}
	(void) fprintf(stderr, "%s: ", Argv[0]);
	perror(stmp);
	Exit(-1);
    }

    /* install signal handler to clean up stmp */
    setup_signals();

    sudoers_fd = open(sudoers, O_RDONLY);
    if (sudoers_fd < 0 && errno != ENOENT) {
	(void) fprintf(stderr, "%s: ", Argv[0]);
	perror(sudoers);
	Exit(-1);
    }

    /*
     * Copy the data
     */
    if (sudoers_fd >= 0) {
	while ((n = read(sudoers_fd, buf, sizeof(buf))) > 0)
	    if (write(stmp_fd, buf, n) != n) {
		(void) fprintf(stderr, "%s: Write failed: ", Argv[0]);
		perror("");
		Exit(-1);
	    }

	(void) close(sudoers_fd);
    }
    (void) close(stmp_fd);

    /*
     * Edit the temp file and parse it (for sanity checking)
     */
    do {
	/*
	 * Build up a buffer to execute
	 */
	if (strlen(Editor) + strlen(stmp) + 30 > sizeof(buf)) {
	    (void) fprintf(stderr, "%s: Buffer too short (line %d).\n",
			   __LINE__, Argv[0]);
	    Exit(-1);
	}
	if (parse_error == TRUE)
	    (void) sprintf(buf, "%s +%d %s", Editor, errorlineno, stmp);
	else
	    (void) sprintf(buf, "%s %s", Editor, stmp);

	/* do the edit -- some SYSV editors return 256 instead of 0 */
	n = system(buf);
	if (n == 0 || n == 256) {
	    struct stat statbuf;	/* for sanity checking */

	    /* make sure stmp exists */
	    if (stat(stmp, &statbuf) < 0) {
		(void) fprintf(stderr,
		    "%s: Can't stat temporary file (%s), %s unchanged.\n",
		    Argv[0], stmp, sudoers);
		Exit(-1);
	    }

	    /* check for zero length file */
	    if (statbuf.st_size == 0) {
		(void) fprintf(stderr,
		    "%s: Zero length temporary file (%s), %s unchanged.\n",
		    Argv[0], stmp, sudoers);
		Exit(-1);
	    }

	    /*
	     * passed sanity checks so reopen stmp file and check
	     * for parse errors.
	     */
	    yyout = stdout;
	    if (parse_error)
		yyin = freopen(stmp, "r", yyin);
	    else
		yyin = fopen(stmp, "r");
	    if (yyin == NULL) {
		(void) fprintf(stderr,
		    "%s: Can't re-open temporary file (%s), %s unchanged.\n",
		    Argv[0], stmp, sudoers);
		Exit(-1);
	    }

	    /* clean slate for each parse */
	    init_parser();

	    /* parse the sudoers file */
	    if (yyparse() && parse_error != TRUE) {
		(void) fprintf(stderr,
		    "%s: Failed to parse temporary file (%s), unknown error.\n",
		    Argv[0], stmp, sudoers);
		parse_error = TRUE;
	    }
	} else {
	    (void) fprintf(stderr, "%s: Editor (%s) failed, %s unchanged.\n",
		Argv[0], Editor, sudoers);
	    Exit(-1);
	}

	/*
	 * Prompt the user for what to do now
	 */
	if (parse_error == TRUE) {
	    switch (whatnow()) {
		case 'q' :	parse_error = FALSE;	/* ignore parse error */
				break;
		case 'x' :	Exit(0);
				break;
	    }
	}
    } while (parse_error == TRUE);

    /*
     * Change mode and ownership of temp file so when
     * we move it to sudoers things are kosher.
     */
    if (chown(stmp, SUDOERS_UID, SUDOERS_GID)) {
	(void) fprintf(stderr,
	    "%s: Unable to set (uid, gid) of %s to (%d, %d): ",
	    Argv[0], stmp, SUDOERS_UID, SUDOERS_GID);
	perror("");
	Exit(-1);
    }
    if (chmod(stmp, SUDOERS_MODE)) {
	(void) fprintf(stderr,
	    "%s: Unable to change mode of %s to %o: ",
	    Argv[0], stmp, SUDOERS_MODE);
	perror("");
	Exit(-1);
    }

    /*
     * Now that we have a sane stmp file (parse ok) it needs to be
     * rename(2)'d to sudoers.  If the rename(2) fails we try using
     * mv(1) in case stmp and sudoers are on different filesystems.
     */
    if (rename(stmp, sudoers))
	if (errno == EXDEV) {
	    char *tmpbuf;

	    (void) fprintf(stderr,
	      "%s: %s and %s not on the same filesystem, using mv to rename.\n",
	      Argv[0], stmp, sudoers);

	    /* Allocate just enough space for tmpbuf */
	    n = sizeof(char) * (strlen(_PATH_MV) + strlen(stmp) +
		  strlen(sudoers) + 4);
	    if ((tmpbuf = (char *) malloc(n)) == NULL) {
		(void) fprintf(stderr,
			      "%s: Cannot alocate memory, %s unchanged: ",
			      Argv[0], sudoers);
		perror("");
		Exit(-1);
	    }

	    /* Build up command and execute it */
	    (void) sprintf(tmpbuf, "%s %s %s", _PATH_MV, stmp, sudoers);
	    if (system(tmpbuf)) {
		(void) fprintf(stderr,
			       "%s: Command failed: '%s', %s unchanged.\n",
			       Argv[0], tmpbuf, sudoers);
		Exit(-1);
	    }
	    (void) free(tmpbuf);
	} else {
	    (void) fprintf(stderr, "%s: Error renaming %s, %s unchanged: ",
				   Argv[0], stmp, sudoers);
	    perror("");
	    Exit(-1);
	}

    return(0);
}


/********************************************************************
 *
 *  dummy *_matches routines
 *
 *  These exist to allow us to use the same parser as sudo(8).
 */

int command_matches(cmnd, user_args, path, sudoers_args)
    char *cmnd;
    char *user_args;
    char *path;
    char *sudoers_args;
{
    return(TRUE);
}


int addr_matches(n)
    char *n;
{
    return(TRUE);
}

int usergr_matches(g, u)
    char *g, *u;
{
    return(TRUE);
}


int netgr_matches(n, h, u)
    char *n, *h, *u;
{
    return(TRUE);
}


/********************************************************************
 *
 *  usage()
 *
 *  Prints a help message and exits w/ exit value of 1.
 */

static void usage()
{
    (void) fprintf(stderr, "usage: %s [-V]\n", Argv[0]);
    Exit(-1);
}


/********************************************************************
 *
 *  Exit()
 *
 *  Unlinks the sudoers temp file (if it exists) and exits.
 *  Used in place of a normal exit() and as a signal handler.
 *  A positive parameter is considered to be a signal and is reported.
 */

static RETSIGTYPE Exit(sig)
    int sig;
{
    (void) unlink(stmp);

    if (sig > 0)
	(void) fprintf(stderr, "%s exiting, caught signal %d.\n", Argv[0], sig);

    exit(-sig);
}


/********************************************************************
 *
 *  whatnow()
 *
 *  Assuming a parse error occurred, prompt the user for what they want
 *  to do now.  Returns first letter of their choice (always lowercase).
 */

static char whatnow()
{
    char choice;
    int ok;

    do {
	ok = FALSE;
	(void) fputs("What now? ", stdout);
	if ((choice = getchar()) != '\n')
	    while (getchar() != '\n')
		;

	if (choice == 'e' || choice == 'x' || choice == 'Q')
	    ok = TRUE;

	/* help message if they gavce us garbage */
	if (!ok)
	    whatnow_help();

    } while (!ok);

    return(choice);
}


/********************************************************************
 *
 *  whatnow_help()
 *
 *  Print out a help message for whatnow().
 */

static void whatnow_help()
{
    (void) printf("Options are:\n");
    (void) printf("  (e)dit sudoers file again\n");
    (void) printf("  e(x)it without saving changes to sudoers file\n");
    (void) printf("  (Q)uit and save changes to sudoers file (DANGER!)\n\n");
}


/********************************************************************
 *
 *  setup_signals()
 *
 *  Install signal handlers for visudo.
 */

static void setup_signals()
{
#ifdef POSIX_SIGNALS
	struct sigaction action;		/* posix signal structure */
#endif /* POSIX_SIGNALS */

	/*
	 * Setup signal handlers
	 */
#ifdef POSIX_SIGNALS
	(void) memset((VOID *)&action, 0, sizeof(action));
	action.sa_handler = Exit;
	action.sa_flags = SA_RESETHAND;
	(void) sigaction(SIGILL, &action, NULL);
	(void) sigaction(SIGTRAP, &action, NULL);
	(void) sigaction(SIGBUS, &action, NULL);
	(void) sigaction(SIGSEGV, &action, NULL);
	(void) sigaction(SIGTERM, &action, NULL);

	action.sa_handler = SIG_IGN;
	action.sa_flags = 0;
	(void) sigaction(SIGHUP, &action, NULL);
	(void) sigaction(SIGINT, &action, NULL);
	(void) sigaction(SIGQUIT, &action, NULL);
#else
	(void) signal(SIGILL, Exit);
	(void) signal(SIGTRAP, Exit);
	(void) signal(SIGBUS, Exit);
	(void) signal(SIGSEGV, Exit);
	(void) signal(SIGTERM, Exit);

	(void) signal(SIGHUP, SIG_IGN);
	(void) signal(SIGINT, SIG_IGN);
	(void) signal(SIGQUIT, SIG_IGN);
#endif /* POSIX_SIGNALS */
}
